xds/internal/xdsclient/clientimpl_watchers.go

```diff
@@ -103,3 +107,80 @@ func (c *clientImpl) WatchEndpoints(clusterName string, cb func(xdsresource.Endp
 		unref()
 	}
 }
+
+// WatchResource uses xDS to discover the resource associated with the provided
+// resource name. The resource type implementation determines how xDS requests
+// are sent out and how responses are deserialized and validated. Upon receipt
+// of a response from the management server, an appropriate callback on the
+// watcher is invoked.
+func (c *clientImpl) WatchResource(rType xdsresource.Type, resourceName string, watcher xdsresource.GenericResourceWatcher) (cancel func()) {
+	// Return early if the client is already closed.
+	//
+	// The client returned from the top-level API is a ref-counted client which
+	// contains a pointer to `clientImpl`. When all references are released, the
+	// ref-counted client sets its pointer to `nil`. And if any watch APIs are
+	// made on such a closed client, we will get here with a `nil` receiver.
+	if c == nil || c.done.HasFired() {
```

Comment:
The ref-counted client `clientRefCounted` embeds a `clientImpl` and the only method it overrides is `Close()` where it handles the reference counts. And the ref-counted client is the singleton. So, when all references are released by repeated calls to its `Close()`, it closes the embedded `clientImpl` and sets the pointer to `nil`. See [here](https://github.com/grpc/grpc-go/blob/56ac86fa0f3940cb79946ce2c6e56f7ee7ecae84/xds/internal/xdsclient/singleton.go#L117). And the reason for setting this pointer to `nil` here is to ensure two things:
- when all references are released, we close the underlying client implementation and clean up resources
- when a subsequent `New` call is made, we create a new `clientImpl`.

Now, we cannot handle the underlying `clientImpl` being `nil` from the ref-counted client without overriding the `WatchResource` method as well. Thinking about it now, it might not be such a bad thing to do, since we have a single `WatchResource` method as opposed to multiple `WatchXxx` methods.

Let me know if you prefer that I do that. If you feel that way, I will most probably add a TODO for that, and take care of it when I get rid of the other `WatchXxx` APIs.

Reply:
This makes sense, thanks.